### PR TITLE
adding Knative Annotation Processor

### PR DESCRIPTION
**Pull request by Uvindu96**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: kubernetes-extension/src/main/resources/META-INF/services/org.ballerinalang.compiler.plugins.CompilerPlugin [hunk 1]
```diff
@@ -1 +1,2 @@
-org.ballerinax.kubernetes.KubernetesPlugin
\ No newline at end of file
+org.ballerinax.kubernetes.KubernetesPlugin
+org.ballerinax.kubernetes.KnativePlugin
```
Comment (suganyasuven) [inline kubernetes-extension/src/main/resources/META-INF/services/org.ballerinalang.compiler.plugins.CompilerPlugin]: Missing a line

File: kubernetes-extension/src/main/java/org/ballerinax/kubernetes/utils/KnativeUtils.java [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+        }
+        try {
+            Files.walk(pathToBeDeleted)
+                    .sorted(Comparator.reverseOrder())
+                    .map(Path::toFile)
+                    .forEach(File::delete);
+        } catch (IOException e) {
+            throw new KubernetesPluginException("unable to delete directory: " + path, e);
+        }
+
```
Comment (suganyasuven) [inline kubernetes-extension/src/main/java/org/ballerinax/kubernetes/utils/KnativeUtils.java]: Shall we remove this line?

File: kubernetes-extension/src/main/java/org/ballerinax/kubernetes/specs/KnativePodSpec.java [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+    /**
+     * Kubernetes service annotations model class.
+     * Knative PodSpec
+     */
+    @JsonProperty("containers")
+    @Valid
+    private List<Container> containers = new ArrayList<Container>();
+
+
```
Comment (suganyasuven) [inline kubernetes-extension/src/main/java/org/ballerinax/kubernetes/specs/KnativePodSpec.java]: Shall we remove this extra line?

File: kubernetes-extension/src/main/java/org/ballerinax/kubernetes/processors/knative/KnativeServiceAnnotationProcessor.java [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+import org.wso2.ballerinalang.compiler.tree.BLangService;
+import org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangExpression;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangListConstructorExpr;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangLiteral;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangRecordLiteral;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangSimpleVarRef;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangTypeInit;
+
+
```
Comment (suganyasuven) [inline kubernetes-extension/src/main/java/org/ballerinax/kubernetes/processors/knative/KnativeServiceAnnotationProcessor.java]: Please remove empty lines at L41, L60, L61, L112, L134, L135, and L398

File: kubernetes-extension/src/main/java/org/ballerinax/kubernetes/processors/knative/KnativeServiceAnnotationProcessor.java [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+    }
+
+    private enum PodTolerationConfiguration {
+        key,
+        operator,
+        value,
+        effect,
+        tolerationSeconds
+    }
+    }
```
Comment (suganyasuven) [inline kubernetes-extension/src/main/java/org/ballerinax/kubernetes/processors/knative/KnativeServiceAnnotationProcessor.java]: Formatting issue

File: kubernetes-extension/src/main/java/org/ballerinax/kubernetes/KnativeArtifactManager.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    private void printInstructions() {
+        printInstruction("");
+        printInstruction("");
+        for (Map.Entry<String, String> instruction : instructions.entrySet()) {
+            printInstruction(instruction.getKey());
+            printInstruction(instruction.getValue());
+            printInstruction("");
+        }
+    }
+    public void populateDeploymentModel() {
```
Comment (suganyasuven) [inline kubernetes-extension/src/main/java/org/ballerinax/kubernetes/KnativeArtifactManager.java]: Add a line after L85 and L100

File: kubernetes-extension/src/main/java/org/ballerinax/kubernetes/KnativePlugin.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+                KnativeAnnotationProcessorFactory.getAnnotationProcessorInstance(annotationKey).processAnnotation
+                        (functionNode, attachmentNode);
+            } catch (KubernetesPluginException e) {
+                dlog.logDiagnostic(Diagnostic.Kind.ERROR, functionNode.getPosition(), e.getMessage());
+            }
+        }
+    }
+
+    @Override
+    public void codeGenerated(PackageID moduleID, Path executableJarFile) {
```
Comment (suganyasuven) [inline kubernetes-extension/src/main/java/org/ballerinax/kubernetes/KnativePlugin.java]: Normally method name starts with a verb

File: kubernetes-extension/src/main/java/org/ballerinax/kubernetes/models/knative/ConfigMapModel.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof ConfigMapModel)) {
+            return false;
+        }
+        ConfigMapModel that = (
+                ConfigMapModel) o;
```
Comment (suganyasuven) [inline kubernetes-extension/src/main/java/org/ballerinax/kubernetes/models/knative/ConfigMapModel.java]: Shall we bring this to the previous line?

File: kubernetes-extension/src/main/java/org/ballerinax/kubernetes/models/knative/ConfigMapModel.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        return Objects.equals(getMountPath(), that.getMountPath());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getMountPath());
+    }
+
```
Comment (kalaiyarasiganeshalingam) [inline kubernetes-extension/src/main/java/org/ballerinax/kubernetes/models/knative/ConfigMapModel.java]: Remove extra lines

File: kubernetes-extension/src/main/java/org/ballerinax/kubernetes/models/knative/EnvVarValueModel.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+                    ", resource='" + resource + '\'' +
+                    '}';
+        }
+    }
+
+    /**
+     * Model for configMapKey type values for environment variables.
+     */
+    public static class ConfigMapKeyValue implements
+            RefValue {
```
Comment (kalaiyarasiganeshalingam) [inline kubernetes-extension/src/main/java/org/ballerinax/kubernetes/models/knative/EnvVarValueModel.java]: Can you move this to before line?

File: kubernetes-extension/src/main/java/org/ballerinax/kubernetes/models/knative/KnativeService.java [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    @JsonProperty("spec")
+    public void setSpec(KnativeServiceSpec spec) {
+        this.spec = spec;
+    }
+
+    public String toString() {
+        return "Deployment(apiVersion=" + this.getApiVersion() + ", kind=" + this.getKind() + ", " +
+                "metadata=" + this.getMetadata() + ", spec=" + this.getSpec() + ")";
+    }
+    public boolean equals(Object o) {
```
Comment (kalaiyarasiganeshalingam) [inline kubernetes-extension/src/main/java/org/ballerinax/kubernetes/models/knative/KnativeService.java]: Can we use the meaningful variable's name?

**Comment by hemikak** [comment]
Merged in https://github.com/ballerinax/kubernetes/pull/448 . Hence closing.

**Pull request closed without merging**